lib/src/main/kotlin/com/swmansion/starknet/crypto/Poseidon.kt

```diff
@@ -0,0 +1,154 @@
+private fun toBigIntegerArray(longArray: LongArray): Array<BigInteger> {
+    val ulongArray = longArray.map { it.toULong() }.toULongArray()
+    return ulongArray.map { BigInteger(it.toString()) }.toTypedArray()
+}
+
+/**
+ * Combine a LongArray of 64-bit chunks into a single BigInteger.
+ *
+ * @param arr the LongArray containing the 64-bit chunks
+ * @return a BigInteger representing the combined value of the input LongArray
+ */
+private fun unsplitBigInteger(arr: LongArray): BigInteger {
+    val powersOfTwo = listOf(
+        BigInteger.valueOf(2).pow(0),
+        BigInteger.valueOf(2).pow(64),
+        BigInteger.valueOf(2).pow(128),
+        BigInteger.valueOf(2).pow(192),
+    )
+    val barr = toBigIntegerArray(arr)
+    // w * 2**0 + x * 2**64 + y * 2**128 + z * 2**192
+    return barr.zip(powersOfTwo).fold(BigInteger.ZERO) { acc, (b, p) -> acc + b * p }
+}
+
+/**
+ * Starknet poseidon utilities.
+ *
+ * Class with utility methods related to starknet poseidon hash calculation.
+ */
+object Poseidon {
+
+    private val m = 3
+    private val r = 2
+
+    init {
+        NativeLoader.load("poseidon_jni")
+    }
+
+    /**
+     * Native hades permutation.
+     */
+    @JvmStatic
+    private external fun hades(
+        values: Array<LongArray>,
+    ): Array<LongArray>
+
+    /**
+     * Compute poseidon hash on single Felt.
+     *
+     * @param x single Felt
+     */
+    @JvmStatic
+    fun poseidonHash(x: Felt): Felt {
+        return unsplitBigInteger(
+            hades(
+                arrayOf(
+                    splitBigInteger(BigInteger(x.decString())),
+                    longArrayOf(0, 0, 0, 0),
+                    longArrayOf(1, 0, 0, 0),
+                ),
+            )[0],
+        ).toFelt
+    }
+
+    /**
+     * Compute poseidon hash on two Felts.
+     *
+     * @param x Felt
+     * @param y Felt
+     */
+    @JvmStatic
+    fun poseidonHash(x: Felt, y: Felt): Felt {
+        return unsplitBigInteger(
+            hades(
+                arrayOf(
+                    splitBigInteger(BigInteger(x.decString())),
+                    splitBigInteger(BigInteger(y.decString())),
+                    longArrayOf(2, 0, 0, 0),
+                ),
+            )[0],
+        ).toFelt
+    }
+
+    /**
+     * Compute poseidon hash on many Felts.
+     *
+     * @param values List of Felts
+     */
+    @JvmStatic
+    fun poseidonHash(values: List<Felt>): Felt {
+        val inputValues = values.toMutableList().apply {
+            add(Felt.ONE)
+            if (size % r == 1) add(Felt.ZERO)
+        }
+
+        var state = Array(m) { longArrayOf(0, 0, 0, 0) }
+
+        for (iter in 0 until inputValues.size step 2) {
+            state = hades(
+                arrayOf(
+                    splitBigInteger(unsplitBigInteger(state[0]) + BigInteger(inputValues[iter].decString())),
+                    splitBigInteger(unsplitBigInteger(state[1]) + BigInteger(inputValues[iter + 1].decString())),
```

Comment:
I'm wondering if maybe these additions should be done in the prime field?

Of course it's a very edge case, and maybe there are some boundaries set in the algorithm itself.

Reply:
The C implementation from submodule specifies the prime field and makes sure it operates within its boundries :+1: